src/path.cr

```diff
@@ -0,0 +1,823 @@
+        name.each_char_with_index do |char, index|
+          case index
+          when 0;
+            if separators.includes?(char)
+              check_for_share = true
+            elsif !char.ascii_letter?
+              return false
+            end
+          when 1;
+            if check_for_share && separators.includes?(char)
+            elsif char == ':'
+              check_for_share = false
+            else
+              return false
+            end
+          else
+            if index == 2 && !check_for_share
+              return separators.includes?(char)
+            end
+            case char
+            when '\\', '/'
+              return false if index == 2 || found_share_name
+              found_share_name = true
+            else
+              return true if found_share_name
+            end
+          end
+        end
+        false
+      else
+        name.starts_with?('/')
+      end
+    end
+
+    def root(name)
+      if windows?
+        if !starts_with_separator?(name) && name[1]? == ':' && separators.includes?(name[2]?)
+          # drive
+          return name[2..2]
+        elsif name.starts_with?("\\\\") || name.starts_with?("//")
+          # UNC share
+          index = 0
+          each_part_separator_index(name) do |pos|
+            if index == 2
+              return name[pos]?.try &.to_s
+            end
+            index += 1
+          end
+          # if path starts with separator but is not UNC share or it is only the share without a root
+          return nil
+        elsif starts_with_separator?(name)
+          return name[0..0]
+        end
+      elsif absolute?(name) # posix
+        return separator_string
+      end
+    end
+
+    def drive(name)
+      if windows?
+        if !starts_with_separator?(name) && name[1]? == ':'
+          return name[0..1]
+        elsif name.starts_with?("\\\\") || name.starts_with?("//")
+          # UNC share
+          index = 0
+          last_pos = 0
+          each_part_separator_index(name) do |pos|
+            if index == 2
+              return name.byte_slice(0, pos)
+            end
+            index += 1
+            last_pos = pos
+          end
+
+          if index == 2 && last_pos < name.bytesize
+            # the entire name is a UNC share
+            return name
+          end
+        end
+      end
+
+      nil
+    end
+  end
+
+  # :nodoc:
+  NATIVE_TYPE = {% if flag?(:windows) %}
+    Type::WINDOWS
+  {% else %}
+    Type::POSIX
+  {% end %}
+
+  # Creates a new `Path` in native flavour.
+  #
+  # When compiling for a windows target, this is equal to `Path.windows()`,
+  # otherwise `Path.posix` is used.
+  def self.new(name : String = "") : Path
+    new(name.check_no_null_byte, NATIVE_TYPE)
+  end
+
+  # ditto
+  def self.new(name : String, *parts) : Path
+    new(name).join(*parts)
+  end
+
+  # ditto
+  def self.new(parts : Array | Tuple) : Path
+    new("").join(parts)
+  end
+
+  # ditto
+  def self.[](*parts) : Path
```

Comment:
These constructors concern me. There's too many and this one seems like an alias.

Reply:
Agree. But I think there should be an easy way for creating Path "literals". 

If we can get the array-like literal `Path{"foo", "bar"}` working properly, that's is probably better (perhaps a bit slower, though).

Reply:
```crystal
require "benchmark"
Benchmark.ips do |bm|
  bm.report("Path{}") { Path{"foo", "bar", "baz", "42"} }
  bm.report("Path[]") { Path["foo", "bar", "baz", "42"] }
end
# Path{} 146.57k (  6.82µs) (±41.01%)  836 B/op   4.42× slower
# Path[] 647.97k (  1.54µs) (±49.08%)  209 B/op        fastest
```

The array-like literal creates an empty instance and calls `#<<` (i.e. `#join`) for each component. The `[]` class method calls `#join` directly with all components, so it joins everything in one go.

Reply:
We should probably have `T{x, y, z}` rewrite to `T.new({x, y, z})` or maybe `T.literal(x, y, z)`. That way you could write an optimal performance. In fact, maybe we could provide both things, depending on what the type has (if it has a `literal` class method we'd use that, otherwise just `new` and <<`). For hashes it's a bit trickier, though.

Reply:
Is may be faster *now* but the idiomatic constructor should be `Path{}` not `Path[]`. We should make `Path{}` faster instead of introducing an alternative that will to be removed. Even if it's slower for the time being.

Reply:
We need to find a consensus for this in #5703

Reply:
I agree with @ysbaddaden . Let's keep only the `Path{}` idiom and we can later improve that story.